Update resolve code to support OpenJDK MH

fengxue-IS opened the pull request:
- Rename methodTypes table to invokeCache table
- new sendResolveOpenJDKInvokeHandle
- new resolveInvokeHandle
- update resolveInvokeDynamic
- update resolveVirtualMethodRefInto
- update resolveStaticMethodRefInto
- Modify Interpreter code to support the new single slot design
- update old_slow_jitResolveHandleMethod

Related: #7352 

Co-authored-by: Babneet Singh <sbabneet@ca.ibm.com>
Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

DanHeidinga commented:
@fengxue-IS can you @ me when this is ready to be reviewed?

DanHeidinga commented:
@fengxue-IS is this ready for another review cycle?

fengxue-IS commented:
@DanHeidinga Apology for the late reply, I have been trying to resolve a crash when testing this change locally enabling OpenJDK_MethodHandle flag. With Tobi & Babneet's help, the problem is because of java code called during memberName resolution and triggered GC which moved the heap object and need to be refetched.

This is ready for review, can you please take a look?

FYI @babsingh @liqunl @nbhuiyan

DanHeidinga commented:
Thanks for the update Jack.  I'm aiming to review this later this week

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+		return result;
 	}
 
+	Trc_VM_resolveInvokeDynamic_Entry(vmThread, callSiteIndex, bsmIndex, resolveFlags);
```
The entry tracepoint should be invoked before the above return statement. The exit tracepoint should also be invoked before the above return statement.

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2084,56 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	if (NULL != result) {
+		return result;
+	} else if (!canRunJavaCode) {
+		return NULL;
```
The code can be rewritten to have only one return statement. This will allow entry and exit tracepoints to be correctly located.

fengxue-IS replied:
fixed

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2084,56 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		}
+
+		if (result != NULL) {
+			/* store result */
```
```suggestion
			/* Store result. */
```

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
-	if (methodHandle != NULL) {
-		return methodHandle;
+	if (result != NULL) {
+		return result;
```
If possible, avoid multiple return statements. This will make it easier to position entry and exit tracepoints.

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+		result = vmThread->currentException;
+	} else if (result == NULL) {
+		setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+	}
```
The below `if (result != NULL)` can be replaced with an `else` block here.
```suggestion
	} else { // (result != NULL)
```

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+	Trc_VM_resolveInvokeDynamic_Resolved(vmThread, callSiteIndex, result);
 
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
 	/* check if an exception is already pending */
```
```suggestion
	/* Check if an exception is already pending. */
```

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
 	/* check if an exception is already pending */
 	if (vmThread->currentException != NULL) {
 		/* Already a pending exception */
```
```suggestion
		/* Already a pending exception. */
```

fengxue-IS replied:
keeping existing format for consistency within the file

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+	}
+
+	if (result != NULL) {
+		/* store result */
```
```suggestion
		/* Store result. */
```

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+		}
+	}
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+	/* check if an exception is already pending */
```
```suggestion
	/* Check if an exception is already pending. */
```

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
+	/* check if an exception is already pending */
+	if (vmThread->currentException != NULL) {
+		/* Already a pending exception */
```
```suggestion
		/* Already a pending exception. */
```

fengxue-IS replied:
keeping existing format for consistency within the file

babsingh commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2105,56 +2260,80 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
+		result = NULL;
+	} else if (result == NULL) {
 		setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
 	}
```
The below `if (result != NULL)` can be replaced with an `else` block here.
```suggestion
	} else { // (result != NULL)
```

babsingh reviewed (changes requested):
A mixture of functional and formatting issues are identified.

fengxue-IS commented on runtime/bcutil/ConstantPoolMap.hpp:
```diff
@@ -287,8 +291,14 @@ class ConstantPoolMap
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+	void markMethodRefAsUsedByInvokeHandle(U_16 methodRefCfrCPIndex)        { mark(methodRefCfrCPIndex, INVOKE_HANDLEEXACT); _invokeCacheCount++; }
+	void markMethodRefAsUsedByInvokeHandleGeneric(U_16 methodRefCfrCPIndex) { mark(methodRefCfrCPIndex, INVOKE_HANDLEGENERIC); _invokeCacheCount++;
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
```
The original code isn't ifdef-ed under any tag, so I considered the code as default.
The `J9VM_OPT_METHOD_HANDLE` flag is not very consistent through the vm code, I think this is mostly due to method handle need to be supported by default since Java7/8

fengxue-IS commented on runtime/vm/callin.cpp:
```diff
@@ -942,6 +942,43 @@ sendForGenericInvoke(J9VMThread *currentThread, j9object_t methodHandle, j9objec
+	Trc_VM_sendResolveOpenJDKInvokeHandle_Entry(currentThread);
+	J9VMEntryLocalStorage newELS;
+	if (buildCallInStackFrame(currentThread, &newELS, true, false)) {
+		/* Convert name and signature to String objects */
```
keeping existing format for consistency within the file

fengxue-IS commented on runtime/vm/callin.cpp:
```diff
@@ -942,6 +942,43 @@ sendForGenericInvoke(J9VMThread *currentThread, j9object_t methodHandle, j9objec
+			j9object_t sigString = mmFuncs->j9gc_createJavaLangString(currentThread, J9UTF8_DATA(sigUTF), J9UTF8_LENGTH(sigUTF), 0);
+			nameString = POP_OBJECT_IN_SPECIAL_FRAME(currentThread);
+			if (NULL != sigString) {
+				/* Run the method */
```
keeping existing format for consistency within the file

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2084,56 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		result = (j9object_t)vmThread->returnValue;
+
+		if (vmThread->currentException != NULL) {
+			/* Already a pending exception */
```
keeping existing format for consistency within the file

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -116,6 +116,40 @@ checkForDecompile(J9VMThread *currentThread, J9ROMMethodRef *romMethodRef, bool
+		if (J9UTF8_LITERAL_EQUALS(methodName, methodNameLength, "linkToInterface")) {
+			isMethodHandle = TRUE;
+		}
+		break;
```
fixed

fengxue-IS commented on runtime/bcutil/ConstantPoolMap.hpp:
```diff
@@ -287,8 +291,14 @@ class ConstantPoolMap
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+	void markMethodRefAsUsedByInvokeHandle(U_16 methodRefCfrCPIndex)        { mark(methodRefCfrCPIndex, INVOKE_HANDLEEXACT); _invokeCacheCount++; }
+	void markMethodRefAsUsedByInvokeHandleGeneric(U_16 methodRefCfrCPIndex) { mark(methodRefCfrCPIndex, INVOKE_HANDLEGENERIC); _invokeCacheCount++;
```
fixed

fengxue-IS commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -2325,6 +2325,25 @@ old_slow_jitResolveHandleMethod(J9VMThread *currentThread)
+	if (NULL == invokeCacheArray) {
+		buildJITResolveFrameWithPC(currentThread, J9_SSF_JIT_RESOLVE_DATA, parmCount, true, 0, jitEIP);
+		// add new resolve code which calls sendResolveInvokeHandle -> MHN.linkMethod()
+		// store the memberName/appendix values in invokeCache[invokeCacheIndex]
```
fixed

fengxue-IS commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8017,23 +8017,22 @@ class INTERPRETER_CLASS
-		if (J9_EXPECTED(NULL != memberName)) {
-			if (J9OBJECT_CLAZZ(_currentThread, memberName) == J9VMJAVALANGINVOKEMEMBERNAME_OR_NULL(_vm)) {
+		if (J9_EXPECTED(NULL != invokeCacheArray)) {
+			if (J9CLASS_IS_ARRAY(J9OBJECT_CLAZZ(_currentThread, invokeCacheArray))) {
```
fixed

fengxue-IS commented on runtime/vm/j9vm.tdf:
```diff
@@ -222,6 +222,13 @@ TraceExit=Trc_VM_resolveStaticFieldRef_Exit Overhead=1 Level=3 Template="resolve
 TraceEntry=Trc_VM_resolveInstanceFieldRef_Entry_old Obsolete Overhead=1 Level=3 Template="resolveInstanceFieldRef(ramCP=%p, cpIndex=%zu, flags=%zx, returnAddress=%p)"
 TraceExit=Trc_VM_resolveInstanceFieldRef_Exit Overhead=1 Level=3 Template="resolveInstanceFieldRef --> result=%d"
 
+TraceEntry=Trc_VM_resolveInvokeHandle_Entry Overhead=1 Level=3 Template="resolveInvokeHandle(ramCP=%p, cpIndex=%zu, flags=%zx)"
```
fixed

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2084,56 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+			setCurrentExceptionUTF(vmThread, J9VMCONSTANTPOOL_JAVALANGNULLPOINTEREXCEPTION, NULL);
+		}
+
+		if (result != NULL) {
```
fixed

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -496,7 +534,47 @@ resolveStaticMethodRefInto(J9VMThread *vmStruct, J9ConstantPool *ramCP, UDATA cp
+	*      - 16 bytes for the original method name ("linkToInterface" is the longest)
+	*  - J9UTF8 for empty signature
+	*/
+	U_8 nameAndNAS[sizeof(J9ROMNameAndSignature) + (sizeof(U_16) + 16) + sizeof(J9UTF8)];
```
```suggestion
	U_8 onStackNAS[sizeof(J9ROMNameAndSignature) + (sizeof(U_16) + 16) + sizeof(J9UTF8)];
```
`nameAndNAS` expands to `nameAndNameAndSiguture` which doesn't make sense.  Better to use the name to indicate why it's special

fengxue-IS replied:
fixed

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -496,7 +534,47 @@ resolveStaticMethodRefInto(J9VMThread *vmStruct, J9ConstantPool *ramCP, UDATA cp
+			memcpy(modifiedMethodName->data, initialMethodName, initialMethodNameLength);
+
+			/* Set flag for partial signature lookup. Signature length is already initialized to 0. */
+			lookupOptions |= J9_LOOK_PARTIAL_SIGNATURE;
```
Have you looked at the `J9_LOOK_DIRECT_NAS` lookupOption?  It allows passing direct pointers which would be much simpler:

Then the `onStackNAS` becomes 
```
U_8 onStackNAS[sizeof(J9NameAndSignature)];
J9UTF8 nullSignature = {0};
```
and the assignment becomes:
```
onStackNAS->name = nameAndSig->name;
onStackNAS->signature = nullSignature;
```

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1368,18 +1446,46 @@ resolveVirtualMethodRefInto(J9VMThread *vmStruct, J9ConstantPool *ramCP, UDATA c
+			U_16 initialMethodNameLength = J9UTF8_LENGTH(nameUTF);
+
+			if (isMethodHandleINL(initialMethodName, initialMethodNameLength)) {
+				J9UTF8 *modifiedMethodName = (J9UTF8 *)(nameAndNAS + sizeof(J9ROMNameAndSignature));
+				J9UTF8 *modifiedMethodSig = (J9UTF8 *)(nameAndNAS + sizeof(nameAndNAS) - sizeof(J9UTF8));
+				memset(nameAndNAS, 0, sizeof(nameAndNAS));
+
+				/* Create new J9ROMNameAndSignature */
+				nameAndSig = (J9ROMNameAndSignature *)nameAndNAS;
+				NNSRP_SET(nameAndSig->name, modifiedMethodName);
+				NNSRP_SET(nameAndSig->signature, modifiedMethodSig);
+
+				modifiedMethodName->length = initialMethodNameLength;
+				memcpy(modifiedMethodName->data, initialMethodName, initialMethodNameLength);
+
+				/* Set flag for partial signature lookup. Signature length is already initialized to 0. */
```
Similar comment applies here as well about using DIRECT option.

Also, names like "modifiedX" should in some way, modify X.  Otherwise the code is confusing

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1611,7 +1716,7 @@ resolveVirtualMethodRefInto(J9VMThread *vmStruct, J9ConstantPool *ramCP, UDATA c
 			}
 		}
-#endif /* J9VM_OPT_METHOD_HANDLE */
+#endif /* defined(J9VM_OPT_METHOD_HANDLE) || defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
```
VM_Assert_

But yes, totally agree!

fengxue-IS replied:
fixed

amicic commented on runtime/gc_structs/ClassIterator.hpp:
```diff
@@ -85,7 +85,11 @@ class GC_ClassIterator {
 		, _constantPoolObjectSlotIterator((J9JavaVM *)env->getLanguageVM(), clazz)
 		, _callSitesIterator(clazz)
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+		, _methodTypesIterator(clazz->romClass->invokeCacheCount, clazz->invokeCache)
```
seems reasonable, but since I don't fully understand relationship between MH and invokeCache,  I'll add @dmitripivkine

amicic replied:
(from my narrow perspective) I'm wondering why the ifdef build flag did not change the name, and should it?

DanHeidinga replied:
The ifdef is for the other all feature which is using the OpenJDK MH implementation.

DanHeidinga replied:
@fengxue-IS Can you open an issue to track renaming the `GC_MethodTypesIterator` to something like `GC_InvokeCacheIterator`?  We need to track it as a cleanup task that should be done before the final cutover but doesn't need to be included in this PR

fengxue-IS replied:
created #11168 to track this cleanup work

liqunl commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -2325,6 +2325,26 @@ old_slow_jitResolveHandleMethod(J9VMThread *currentThread)
+		}
+		goto retry;
+	}
+	JIT_RETURN_UDATA(invokeCacheArray);
```
JIT expects this helper to return the address of the entry, not the content of the entry. See the current implementation.

fengxue-IS replied:
fixed

DanHeidinga commented:
@fengxue-IS ping me when the review has been addressed and I'll give this another look

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2084,56 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		return NULL;
+	}
+
+	Trc_VM_resolveInvokeHandle_Entry(vmThread, ramCP, cpIndex, resolveFlags);
```
fixed

fengxue-IS commented on runtime/bcutil/ConstantPoolMap.hpp:
```diff
@@ -287,8 +291,14 @@ class ConstantPoolMap
+#else /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
 	void markMethodRefAsUsedByInvokeHandle(U_16 methodRefCfrCPIndex)        { mark(methodRefCfrCPIndex, INVOKE_HANDLEEXACT); _methodTypeCount++; }
 	void markMethodRefAsUsedByInvokeHandleGeneric(U_16 methodRefCfrCPIndex) { mark(methodRefCfrCPIndex, INVOKE_HANDLEGENERIC); _methodTypeCount++; }
+#endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
```
fixed

fengxue-IS commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -2325,6 +2325,26 @@ old_slow_jitResolveHandleMethod(J9VMThread *currentThread)
+		/* add new resolve code which calls sendResolveInvokeHandle -> MHN.linkMethod()
+		 * store the memberName/appendix values in invokeCache[invokeCacheIndex]
+		 */
+		vm->internalVMFunctions->resolveInvokeHandle(currentThread, ramConstantPool, cpIndex, J9_RESOLVE_FLAG_RUNTIME_RESOLVE);
```
the `targetClass` is resolved as part of the `resolveInvokeHandle` call

fengxue-IS commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -4559,8 +4575,10 @@ typedef struct J9InternalVMFunctions {
 	void  (JNICALL *sendForGenericInvoke)(struct J9VMThread *vmThread, j9object_t methodHandle, j9object_t methodType, UDATA dropFirstArg) ;
 	void  (JNICALL *jitFillOSRBuffer)(struct J9VMThread *vmContext, void *osrBlock) ;
 	void  (JNICALL *sendResolveMethodHandle)(struct J9VMThread *vmThread, UDATA cpIndex, J9ConstantPool *ramCP, J9Class *definingClass, J9ROMNameAndSignature* nameAndSig) ;
+	j9object_t ( *resolveInvokeHandle)(struct J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpIndex, UDATA resolveFlags) ;
```
fixed naming to both include `OpenJDK`

fengxue-IS commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -8017,23 +8017,23 @@ class INTERPRETER_CLASS
-			if (J9OBJECT_CLAZZ(_currentThread, memberName) == J9VMJAVALANGINVOKEMEMBERNAME_OR_NULL(_vm)) {
+		if (J9_EXPECTED(NULL != invokeCacheArray)) {
+			J9Class *clazz = J9OBJECT_CLAZZ(_currentThread, invokeCacheArray);
+			if (J9CLASS_IS_ARRAY(clazz)) {
```
added comments

fengxue-IS commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2086,54 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+		}
+	}
+	Trc_VM_resolveInvokeHandle_Exit(vmThread, result);
+#endif /* defined(J9VM_OPT_OPENJDK_METHODHANDLE) */
```
added `Trc_VM_Assert_ShouldNeverHappen`

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1611,7 +1700,9 @@ resolveVirtualMethodRefInto(J9VMThread *vmStruct, J9ConstantPool *ramCP, UDATA c
 		}
-#endif /* J9VM_OPT_METHOD_HANDLE */
+#else /* defined(J9VM_OPT_METHOD_HANDLE) */
+	Trc_VM_Assert_ShouldNeverHappen();
```
Does this need to be indented 1 more level to match the surrounding code?
```suggestion
		Trc_VM_Assert_ShouldNeverHappen();
```

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -1979,6 +2070,55 @@ resolveMethodHandleRef(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+	UDATA invokeCacheIndex = ramCPEntry->methodIndexAndArgCount >> 8;
+	J9Class *ramClass = J9_CLASS_FROM_CP(ramCP);
+	j9object_t *invokeCache = ramClass->invokeCache + invokeCacheIndex;
+	result = *invokeCache;
```
```suggestion
	j9object_t result = *invokeCache;
```

DanHeidinga replied:
Declare the local here and it only has 1 value - no need to reason about where the original NULL leaks to

DanHeidinga commented on runtime/vm/resolvesupport.cpp:
```diff
@@ -2100,61 +2240,81 @@ resolveConstantDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA cpInde
+					result = *callSite;
+				}
+			}
+		}
```
Is there a `}` missing here?  The indentation doesn't match the `if (NULL == result) {` statement.

fengxue-IS commented:
- Fixed formating/comments as suggested in review
- rename `resolveInvokeHandle` to `resolveOpenJDKInvokeHandle` for consistency
- new `Trc_VM_Assert_ShouldNeverHappen` for unreachable code path
- added writeBarrier before CAS on the result array to ensure array elements are written before the array can be accessed
- modified method lookup to use `J9_LOOK_DIRECT_NAS` flag

@DanHeidinga can you please take another look?

DanHeidinga commented:
@liqunl are you OK with the JIT helper change?

DanHeidinga commented:
Jenkins test sanity zlinux jdk8

DanHeidinga commented:
Jenkins test sanity win jdk11

DanHeidinga commented:
Jenkins test sanity osx jdk15

fengxue-IS commented:
Fixed compiler warning about unused variable `nullSignature`, local linux Java 11 compile passed.
@DanHeidinga can you restart the pr build

DanHeidinga commented:
Jenkins test sanity osx jdk15

DanHeidinga commented:
Jenkins test sanity win jdk11

DanHeidinga commented:
Jenkins test sanity zlinux jdk8

DanHeidinga commented:
Note, all the builds have passed now.  I'm ready to merge this once the conflict is resolved.

Any remaining concerns @babsingh @liqunl?

@fengxue-IS can you resolve the conflict in jvm.tdf?

babsingh commented:
No more concerns.

fengxue-IS commented:
Fixed conflict in j9vm.tdf, local compile passed
